### PR TITLE
feat(EMS-682-634): Phone number validation

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Changes made:
* Added shared validation for phone numbers using google-libphonenumber package
* Added tests
* Improved testing

### CODE REVIEW

File: e2e-tests/constants/website-address-examples.js [hunk 1]
```diff
@@ -0,0 +1 @@
+export const VALID_WEBSITE = 'www.google.com';
```
Comment (ttbarnes) [inline e2e-tests/constants/website-address-examples.js:1]: Do you think we should have this structure?
```js
constants/examples/phone-number
constants/examples/website-address
```

Then could export as `LONG`, `INTERNATIONAL`, `VALID`?

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-phone-number-validation.spec.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    it('should not display validation errors', () => {
+      companyDetails.companiesHouseSearch().clear().type(COMPANIES_HOUSE_NUMBER);
+      yesRadioInput().eq(0).click();
+      yesRadioInput().eq(1).click();
+      companyDetails.companyWebsite().clear().type(VALID_WEBSITE);
+      companyDetails.phoneNumber().clear().type(VALID_PHONE_NUMBER);
+      submitButton().click();
+      partials.errorSummaryListItems().should('have.length', 0);
+    });
+  });
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-phone-number-validation.spec.js:126]: i thought there was a lot more possible phone numbers listed in figma, should we be testing the possibilities? Also do we need to test mobile number?

File: src/ui/server/controllers/insurance/business/company-details/validation/company-details/rules/phone-number.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * validates phone number input is the correct format
+ * @param {RequestBody} responseBody
+ * @param {object} errors
+ * @returns {object} errors
+ */
+const phoneNumber = (responseBody: RequestBody, errors: object) => {
+  let updatedErrors = errors;
+
+  // as field is optional, only validate if it is not an empty string
+  if (responseBody[PHONE_NUMBER]) {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/validation/company-details/rules/phone-number.ts:22]: Could use `objectHasProperty` helper function here

File: src/ui/server/shared-validation/phone-number/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    });
+
+    it('should not display error if valid mobile number', () => {
+      const phoneNumber = '07458045200';
+
+      const result = validatePhoneNumber(phoneNumber, fieldId, errorMessage, errors);
+
+      expect(result).toEqual({});
+    });
+  });
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/phone-number/index.test.ts:98]: Same comment as in E2E tests, seems there are more possibilities that we should test?

File: src/ui/server/shared-validation/phone-number/index.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * @param {string} fieldId
+ * @param {string} errorMessage
+ * @param {object} errors
+ * @returns {object} updatedErrors
+ */
+const validatePhoneNumber = (phoneNumber: string, fieldId: string, errorMessage: string, errors: object) => {
+  let updatedErrors = errors;
+
+  // if not valid for UK, then returns validation error
+  if (!phoneUtil.isValidNumberForRegion(phoneUtil.parse(phoneNumber, defaultCountryUK), defaultCountryUK)) {
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/phone-number/index.ts:20]: Can we uppercase the constant please to be consistent with other constants

Comment (Zainzzkk) [reply src/ui/server/shared-validation/phone-number/index.ts:20]: done

File: e2e-tests/constants/website-address-examples.js (hunk 1 above)
Comment (Zainzzkk) [reply e2e-tests/constants/website-address-examples.js:1]: done

File: src/ui/server/shared-validation/phone-number/index.test.ts (hunk 4 above)
Comment (Zainzzkk) [reply src/ui/server/shared-validation/phone-number/index.test.ts:98]: done

File: src/ui/server/controllers/insurance/business/company-details/validation/company-details/rules/phone-number.ts (hunk 3 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/business/company-details/validation/company-details/rules/phone-number.ts:22]: done

File: e2e-tests/constants/examples/phone-number-examples.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+export const INVALID_PHONE_NUMBERS = {
+  LONG_LANDLINE: '020727180101',
```
Comment (ttbarnes) [inline e2e-tests/constants/examples/phone-number-examples.js:2]: Very minor but could we have `LANDLINE` first, like the others

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-invalid-phone-number-validation.spec.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        submitButton().click();
+        partials.errorSummaryListItems().should('have.length', 1);
+        partials.errorSummaryListItems().first().invoke('text')
+          .then((text) => {
+            expect(text.trim()).equal(errorMessage);
+          });
+        companyDetails.phoneNumberError().invoke('text')
+          .then((text) => {
+            expect(text.trim()).equal(`Error: ${errorMessage}`);
+          });
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-invalid-phone-number-validation.spec.js:200]: Finding these tests a bit hard to read, could we add some spacing? Also could we extract the first few lines of each test? Seems to be repeated a lot, something like this?

```js
const completeAllOtherFields = () => {
  companyDetails.companiesHouseSearch().clear().type(COMPANIES_HOUSE_NUMBER);
  yesRadioInput().eq(0).click();
  yesRadioInput().eq(1).click();

  companyDetails.companyWebsite().clear().type(WEBSITE_EXAMPLES.VALID);  
};


it('should...', () => {
  completeAllOtherFields();

  companyDetails.phoneNumber().clear().type(INVALID_PHONE_NUMBERS.LANDLINE_LETTER);
  submitButton().click();

  partials.errorSummaryListItems().should('have.length', 1);
  partials.errorSummaryListItems().first().invoke('text')
    .then((text) => {
      expect(text.trim()).equal(errorMessage);
    });

  companyDetails.phoneNumberError().invoke('text')
    .then((text) => {
      expect(text.trim()).equal(`Error: ${errorMessage}`);
    });
});
```

File: src/ui/server/controllers/insurance/business/company-details/index.ts [hunk 8]
```diff
@@ -195,15 +197,7 @@ const post = async (req: Request, res: Response) => {
...
-    return res.render(TEMPLATE, {
-      ...insuranceCorePageVariables({
-        PAGE_CONTENT_STRINGS: COMPANY_DETAILS,
-        BACK_LINK: req.headers.referer,
-      }),
-      ...pageVariables(application.referenceNumber),
-      validationErrors,
-      submittedValues,
-    });
+    return res.redirect(NATURE_OF_BUSINESS);
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.ts:200]: Can remove the TODO comment now 🎉

File: e2e-tests/constants/examples/phone-number-examples.js (hunk 6 above)
Comment (Zainzzkk) [reply e2e-tests/constants/examples/phone-number-examples.js:2]: done

File: src/ui/server/controllers/insurance/business/company-details/index.ts (hunk 8 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/business/company-details/index.ts:200]: done

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-invalid-phone-number-validation.spec.js (hunk 7 above)
Comment (Zainzzkk) [reply e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/validation/company-details-invalid-phone-number-validation.spec.js:180]: done

**Review by ttbarnes — APPROVED** [review]
Great work! 🚀

**Pull request merged**